core/src/main/scala/kafka/server/KafkaConfig.scala

```diff
@@ -560,7 +560,17 @@ class KafkaConfig private(doLog: Boolean, val props: util.Map[_, _])
     controllerListenerNames.flatMap { name =>
       controllerAdvertisedListeners
         .find(endpoint => endpoint.listenerName.equals(ListenerName.normalised(name)))
-        .orElse(controllerListenersValue.find(endpoint => endpoint.listenerName.equals(ListenerName.normalised(name))))
+        .orElse(
+          // If users don't define advertised.listeners, the advertised controller listeners inherit from listeners configuration
+          // which match listener names in controller.listener.names. Also, removing "0.0.0.0" host to avoid validation errors.
+          controllerListenersValue
+            .find(endpoint => endpoint.listenerName.equals(ListenerName.normalised(name)))
+            .map(endpoint => if (endpoint.host == "0.0.0.0") {
+              new EndPoint(null, endpoint.port, endpoint.listenerName, endpoint.securityProtocol)
```

Comment:
I think that `null` or `""` works in some network because in the `ControllerServer` code we do a reverse lookup.
```scala
        val listenerInfo = ListenerInfo
          .create(config.effectiveAdvertisedControllerListeners.map(_.toJava).asJava)
          .withWildcardHostnamesResolved()
          .withEphemeralPortsCorrected(name => socketServer.boundPort(new ListenerName(name)))
```
Notice the use of `withWildcardHostnamesResolved`. Can you document this in the comment above?

I should also mentioned that `0.0.0.0` is IPv4. Kafka also supports IPv6 which is encoded as `::`.